# Exploring eBay Car Sale Data

This project shall work with a dataset of used cars from eBay Kleinanzeigen, a classifieds section of the German eBay website.

The dataset was originally scraped and uploaded to Kaggle by user [orgesleka](https://www.kaggle.com/orgesleka). The original dataset isn't available on Kaggle anymore, but you can find it [here](https://data.world/data-society/used-cars-data).

One modification has been made from the original dataset:

- 50,000 data points have been sampled from the full dataset.

The data dictionary provided with the data is as follows:

- `dateCrawled`: When this ad was first crawled. All field values are taken from this date.

- `name`: Name of the car.

- `seller`: Whether the seller is private or a dealer.

- `OfferType`: Whether the seller is private or a dealer.

- `price`: The price on the ad to sell the car.

- `abtest`: Whether the listing is included in an A/B test.

- `vehicleType`: The vehicle Type.

- `yearOfRegistration`: The year in which the car was first registered.

- `gearbox`: The transmission type.

- `powerPS`: The power of the car in PS.

- `model`: The car model name.

- `kilometer`: How many kilometers the car has driven.

- `monthOfRegistration`: The month in which the car was first registered.

- `fuelType`: What type of fuel the car uses.

- `brand`: The brand of the car.

- `notRepairedDamage`: If the car has a damage wich is not yet repaired.

- `dateCreated`: The date on which the eBay listing was created.

- `nrOfPictures`: The number of pictures in the ad.

- `postalCode`: The postal code for the location of the vehicle.

- `lastSeenOnLine`: When the crawler saw this ad last online.

The aim of this project is to clean the data and analyse the included used car listings. We shall now import the libraries we need and read the dataset into pandas.

In [1]:
# import the pandas library and read "autos.csv" in pandas
import pandas as pd
autos = pd.read_csv("autos.csv", encoding="Latin-1")

In [36]:
# display information about the data set and print first six rows
autos.info()
autos.head(6)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

dateCrawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
5  2016-03-21 13:47:45  Chrysler_Grand_Voyager_2.8_CRD_Aut.Limited_Sto...   

   seller offerType   price   abtest vehicleType  yearOfRegistration  \
0  privat   Angebot  $5,000  control         bus                2004   
1  privat   Angebot  $8,500  control   limousine                1997   
2  privat   Angebot  $8,990     test   limousine                2009   
3  privat   Angebot  $4,350  control  kleinwagen                2007   
4  privat   Angebot  $1,350     test       kombi                2003   
5  privat   Angebot  $7,900     test         bus                2006   

     gearbox  powerPS    model   odometer  monthOfRegistration fuelType  \
0    manuell      158   andere  150,000km                    3      lpg   
1  automatik      286      7er  150,000km                    6   benzin   
2    manuell      102     golf   70,000km                    7   benzin   
3  automatik       71   fortwo   70,000km                    6   benzin   
4    manuell        0    focus  150,000km                    7   benzin   
5  automatik      150  voyager  150,000km                    4   diesel   

        brand notRepairedDamage          dateCreated  nrOfPictures  \
0     peugeot              nein  2016-03-26 00:00:00             0   
1         bmw              nein  2016-04-04 00:00:00             0   
2  volkswagen              nein  2016-03-26 00:00:00             0   
3       smart              nein  2016-03-12 00:00:00             0   
4        ford              nein  2016-04-01 00:00:00             0   
5    chrysler               NaN  2016-03-21 00:00:00             0   

   postalCode             lastSeen  
0       79588  2016-04-06 06:45:54  
1       71034  2016-04-06 14:45:08  
2       35394  2016-04-06 20:15:37  
3       33729  2016-03-15 03:16:28  
4       39218  2016-04-01 14:38:50  
5       22962  2016-04-06 09:45:21

The dataset we're looking into has 20 columns and 50000 rows. Only 5 out of the 20 columns have a dtype `int64`, while the rest are dtype `object`. 

Some columns have null values, but none have more than ~20% null values.

## Cleaning Column Names
The column names use camelcase instead of Python's preferred snakecase, which means we can't just replace the spaces with underscores. 

We shall now convert the column names from camelcase to snakecase and reword some of the column names based on the data dictionary to be more descriptive.

In [37]:
# column names in the dataset
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [38]:
# edit column names
autos.rename({"yearOfRegistration": "registration_year",
              "monthOfRegistration": "registration_month",
              "notRepairedDamage": "unrepaired_damage",
              "dateCreated": "ad_created",
              "dateCrawled": "date_crawled",
              "offerType": "offer_type",
              "vehicleType": "vehicle_type",
              "powerPS": "power_ps",
              "fuelType": "fuel_type",
              "nrOfPictures": "nr_of_pictures",
              "postalCode": "postal_code",
              "lastSeen": "last_seen"}, axis=1, inplace=True
              )

# display first five rows
autos.head()

date_crawled                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offer_type   price   abtest vehicle_type  registration_year  \
0  privat    Angebot  $5,000  control          bus               2004   
1  privat    Angebot  $8,500  control    limousine               1997   
2  privat    Angebot  $8,990     test    limousine               2009   
3  privat    Angebot  $4,350  control   kleinwagen               2007   
4  privat    Angebot  $1,350     test        kombi               2003   

     gearbox  power_ps   model   odometer  registration_month fuel_type  \
0    manuell       158  andere  150,000km                   3       lpg   
1  automatik       286     7er  150,000km                   6    benzin   
2    manuell       102    golf   70,000km                   7    benzin   
3  automatik        71  fortwo   70,000km                   6    benzin   
4    manuell         0   focus  150,000km                   7    benzin   

        brand unrepaired_damage           ad_created  nr_of_pictures  \
0     peugeot              nein  2016-03-26 00:00:00               0   
1         bmw              nein  2016-04-04 00:00:00               0   
2  volkswagen              nein  2016-03-26 00:00:00               0   
3       smart              nein  2016-03-12 00:00:00               0   
4        ford              nein  2016-04-01 00:00:00               0   

   postal_code            last_seen  
0        79588  2016-04-06 06:45:54  
1        71034  2016-04-06 14:45:08  
2        35394  2016-04-06 20:15:37  
3        33729  2016-03-15 03:16:28  
4        39218  2016-04-01 14:38:50

## Initial Exploration and Cleaning

Now we shall do some basic data exploration to determine what other cleaning tasks need to be done. Initially we will look for:

- Text columns where all or almost all values are the same. These can often be dropped as they don't have useful information for analysis.
- Examples of numeric data stored as text which can be cleaned and converted.

The following methods are helpful for exploring the data:
- `DataFrame.describe()`(with `include='all'` to get both categorical and numerical columns).
- `Series.value_counts()` and `Series.head()` if any columns need a closer look.

In [39]:
# describe the entire autos data set
autos.describe(include='all')

date_crawled         name  seller offer_type  price abtest  \
count                 50000        50000   50000      50000  50000  50000   
unique                48213        38754       2          2   2357      2   
top     2016-04-02 15:49:30  Ford_Fiesta  privat    Angebot     $0   test   
freq                      3           78   49999      49999   1421  25756   
mean                    NaN          NaN     NaN        NaN    NaN    NaN   
std                     NaN          NaN     NaN        NaN    NaN    NaN   
min                     NaN          NaN     NaN        NaN    NaN    NaN   
25%                     NaN          NaN     NaN        NaN    NaN    NaN   
50%                     NaN          NaN     NaN        NaN    NaN    NaN   
75%                     NaN          NaN     NaN        NaN    NaN    NaN   
max                     NaN          NaN     NaN        NaN    NaN    NaN   

       vehicle_type  registration_year  gearbox      power_ps  model  \
count         44905       50000.000000    47320  50000.000000  47242   
unique            8                NaN        2           NaN    245   
top       limousine                NaN  manuell           NaN   golf   
freq          12859                NaN    36993           NaN   4024   
mean            NaN        2005.073280      NaN    116.355920    NaN   
std             NaN         105.712813      NaN    209.216627    NaN   
min             NaN        1000.000000      NaN      0.000000    NaN   
25%             NaN        1999.000000      NaN     70.000000    NaN   
50%             NaN        2003.000000      NaN    105.000000    NaN   
75%             NaN        2008.000000      NaN    150.000000    NaN   
max             NaN        9999.000000      NaN  17700.000000    NaN   

         odometer  registration_month fuel_type       brand unrepaired_damage  \
count       50000        50000.000000     45518       50000             40171   
unique         13                 NaN         7          40                 2   
top     150,000km                 NaN    benzin  volkswagen              nein   
freq        32424                 NaN     30107       10687             35232   
mean          NaN            5.723360       NaN         NaN               NaN   
std           NaN            3.711984       NaN         NaN               NaN   
min           NaN            0.000000       NaN         NaN               NaN   
25%           NaN            3.000000       NaN         NaN               NaN   
50%           NaN            6.000000       NaN         NaN               NaN   
75%           NaN            9.000000       NaN         NaN               NaN   
max           NaN           12.000000       NaN         NaN               NaN   

                 ad_created  nr_of_pictures   postal_code            last_seen  
count                 50000         50000.0  50000.000000                50000  
unique                   76             NaN           NaN                39481  
top     2016-04-03 00:00:00             NaN           NaN  2016-04-07 06:17:27  
freq                   1946             NaN           NaN                    8  
mean                    NaN             0.0  50813.627300                  NaN  
std                     NaN             0.0  25779.747957                  NaN  
min                     NaN             0.0   1067.000000                  NaN  
25%                     NaN             0.0  30451.000000                  NaN  
50%                     NaN             0.0  49577.000000                  NaN  
75%                     NaN             0.0  71540.000000                  NaN  
max                     NaN             0.0  99998.000000                  NaN

In the `autos` dataset there are two columns, `seller` and `offer_type`, which have mostly one value that are candidates to be dropped.

The `seller` column has 49999 values as `privat` and only one as `gewerblich`, i.e. 49999 sellers are private individuals and only one is a commercial establishment.

The `offer_type` column, on the other hand, has 49999 values as `Angebot` which is offer in German, and only one value is `Gesuch`, which is request in German.

Lastly, the `nr_of_pictures` column countains only one value `0`.

We shall now drop columns that mostly have one value and are useless for our analysis.

In [40]:
# drop unnecessary columns
autos.drop(columns=["seller", "offer_type", "nr_of_pictures"],
          inplace=True
          )

The `price` and `odometer` columns are numeric values stored as text so, firstly, the non-numeric charcters will be removed, then the column shall be converted to a numeric dtype and, finally, `DataFrame.rename()` shall be used to rename the `odometer` column to `odometer_km` and the `price` column to `price_dollars`.

In [41]:
# remove non-numeric characters and convert to numeric dtype
autos["price"] = (autos["price"].str.replace("$", "")
                               .str.replace(",","")
                               .astype(float)
                 )
                  
autos["odometer"] = (autos["odometer"].str.replace("km", "")
                                     .str.replace(",","")
                                     .astype(float)
                    )

# rename columns
autos.rename({"odometer": "odometer_km",
              "price":"price_dollars"},
              axis=1,
              inplace=True
            )

We shall continue exploring the data, specifically looking for data that doesn't look right. We shall start by analysing the `odometer_km` and `price` columns. The steps we shall take are as follows:

- Analyse the columns using minimum and maximum values and look for any values that look unrealistically high or low (outliers) that we might want to remove.

- We shall use:

1. `Series.unique().shape` to see how many uique values
2. `Series.describe()` to view min/max/median/mean etc.
3. `Series.value_counts()`, with some variations:
  - chanined to `.head()` if there are lots of values.
  - Because `Series.value_counts()` returns a series we can use `Series.sort_index()` with `ascending=``True`or`False` to view the highest and lowest values with their counts (can also chain to `head` here.
4. When removing outliers, we can do `df[(df["col"] >= x ) & (df["col"] <= y )]`, but it's more readable to use `df[df["col"].between(x,y)]`

## Exploring the `odometer_km` column

In [42]:
# find total number of values in "odometer_km" column
autos["odometer_km"].unique().shape

(13,)

The `autos["odometer_km"]` has 13 values. We shall now examine the highest and lowest values in the column.

In [43]:
# describe "odometer_km" values
autos["odometer_km"].describe()

count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64

The highest and lowest values in the `autos["odometer_km"]` column are `150000.000000` and `5000.000000` respectively, and neither are unusual with respect to the `autos["odometer_km"]` column.

The mean and median values in the column are `125732.700000` and `150000.000000` respectively.

We shall further explore this column.

In [44]:
# display all values in a descending order
autos["odometer_km"].value_counts().sort_index(ascending=False)

150000.0    32424
125000.0     5170
100000.0     2169
90000.0      1757
80000.0      1436
70000.0      1230
60000.0      1164
50000.0      1027
40000.0       819
30000.0       789
20000.0       784
10000.0       264
5000.0        967
Name: odometer_km, dtype: int64

## Exploring the `price_dollars` column

To summarise, none of the values in the column seem to be, in any way, unusual. We shall now examine the `autos["price_dollars"]` column and see if we can find any irregularities.

In [45]:
# find total number of values in "price_dollars" column
autos["price_dollars"].unique().shape

(2357,)

There are far more values values in the `autos["price_dollars"]` column, i.e. 2357, than there were in the `autos["odometer_km"]` column.

In [46]:
# describe "price_dollars" column
autos["price_dollars"].describe()

count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price_dollars, dtype: float64

Both the minimum and maximum values in this column seem rather unusual. `0.000000e+00` and `1.000000e+08` are too low and too high prices for used cars. We shall explore the data further.

In [47]:
# determine top 25 prices
autos["price_dollars"].value_counts().sort_index(ascending=False).head(25)

99999999.0    1
27322222.0    1
12345678.0    3
11111111.0    2
10000000.0    1
3890000.0     1
1300000.0     1
1234566.0     1
999999.0      2
999990.0      1
350000.0      1
345000.0      1
299000.0      1
295000.0      1
265000.0      1
259000.0      1
250000.0      1
220000.0      1
198000.0      1
197000.0      1
194000.0      1
190000.0      1
180000.0      1
175000.0      1
169999.0      1
Name: price_dollars, dtype: int64

All the prices above `350000.0` apppear to be random key entries or erroneous data and hence shall be considered outliers.

In [48]:
# determine bottom 25 prices
autos["price_dollars"].value_counts().sort_index(ascending=True).head(25)

0.0     1421
1.0      156
2.0        3
3.0        1
5.0        2
8.0        1
9.0        1
10.0       7
11.0       2
12.0       3
13.0       2
14.0       1
15.0       2
17.0       3
18.0       1
20.0       4
25.0       5
29.0       1
30.0       7
35.0       1
40.0       6
45.0       4
47.0       1
49.0       4
50.0      49
Name: price_dollars, dtype: int64

`1421` entries for `0.0` can't be an accident, but for convinience's sake we shall consider them as outliers and remove them from our data.

In [49]:
# remove erroneous data and updating autos
autos = autos[autos["price_dollars"].between(1.0, 350000.0)]

## Exploring the date columns

In [50]:
date_crawled = autos["date_crawled"].str[:10]
date_crawled.value_counts(normalize=True, dropna=False)

2016-04-03    0.038608
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-12    0.036920
2016-03-14    0.036549
2016-04-04    0.036487
2016-03-07    0.036014
2016-04-02    0.035478
2016-03-28    0.034860
2016-03-19    0.034778
2016-03-15    0.034284
2016-03-29    0.034099
2016-03-30    0.033687
2016-04-01    0.033687
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-22    0.032987
2016-03-11    0.032575
2016-03-23    0.032225
2016-03-26    0.032204
2016-03-10    0.032184
2016-03-31    0.031834
2016-03-17    0.031628
2016-03-25    0.031607
2016-03-27    0.031092
2016-03-16    0.029610
2016-03-24    0.029342
2016-03-05    0.025327
2016-03-13    0.015670
2016-03-06    0.014043
2016-04-05    0.013096
2016-03-18    0.012911
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_crawled, dtype: float64

The highest data crawling, as can be seen above, has been done on `2016-04-03`. Let's explore the data further. 

In [51]:
date_crawled.value_counts().sort_index()

2016-03-05    1230
2016-03-06     682
2016-03-07    1749
2016-03-08    1617
2016-03-09    1607
2016-03-10    1563
2016-03-11    1582
2016-03-12    1793
2016-03-13     761
2016-03-14    1775
2016-03-15    1665
2016-03-16    1438
2016-03-17    1536
2016-03-18     627
2016-03-19    1689
2016-03-20    1840
2016-03-21    1815
2016-03-22    1602
2016-03-23    1565
2016-03-24    1425
2016-03-25    1535
2016-03-26    1564
2016-03-27    1510
2016-03-28    1693
2016-03-29    1656
2016-03-30    1636
2016-03-31    1546
2016-04-01    1636
2016-04-02    1723
2016-04-03    1875
2016-04-04    1772
2016-04-05     636
2016-04-06     154
2016-04-07      68
Name: date_crawled, dtype: int64

The time frame for data crawling has been from `2016-03-05` to `2016-04-07`. We shall explore the `autos["ad_created"]` column in a similar way.

In [52]:
ad_created = autos["ad_created"].str[:10]
print(ad_created.value_counts(normalize=True, dropna=False))
print('\n')
print(ad_created.value_counts())

2016-04-03    0.038855
2016-03-20    0.037949
2016-03-21    0.037579
2016-04-04    0.036858
2016-03-12    0.036755
                ...   
2015-11-10    0.000021
2015-12-30    0.000021
2016-02-01    0.000021
2015-09-09    0.000021
2016-01-22    0.000021
Name: ad_created, Length: 76, dtype: float64


2016-04-03    1887
2016-03-20    1843
2016-03-21    1825
2016-04-04    1790
2016-03-12    1785
              ... 
2015-11-10       1
2015-12-30       1
2016-02-01       1
2015-09-09       1
2016-01-22       1
Name: ad_created, Length: 76, dtype: int64


On `2016-04-03`, the highest number of ad listings, i.e. 1887 ads were created. 

In [53]:
ad_created.value_counts().sort_index()

2015-06-11       1
2015-08-10       1
2015-09-09       1
2015-11-10       1
2015-12-05       1
              ... 
2016-04-03    1887
2016-04-04    1790
2016-04-05     574
2016-04-06     158
2016-04-07      61
Name: ad_created, Length: 76, dtype: int64

Ad listings for used cars were created in the time frame starting at `2015-06-11` and ending at `2016-04-07`.

We shall now look into the `autos["last_seen"]` column of the `autos` data set.

In [54]:
last_seen = autos["last_seen"].str[:10]
last_seen.value_counts()

2016-04-06    10772
2016-04-07     6408
2016-04-05     6059
2016-03-17     1364
2016-04-03     1224
2016-04-02     1210
2016-03-30     1203
2016-04-04     1189
2016-03-31     1155
2016-03-12     1155
2016-04-01     1107
2016-03-29     1085
2016-03-22     1038
2016-03-28     1013
2016-03-20     1003
2016-03-21     1002
2016-03-24      960
2016-03-25      933
2016-03-23      900
2016-03-26      816
2016-03-16      799
2016-03-15      771
2016-03-19      769
2016-03-27      760
2016-03-14      612
2016-03-11      601
2016-03-10      518
2016-03-09      466
2016-03-13      432
2016-03-08      360
2016-03-18      357
2016-03-07      262
2016-03-06      210
2016-03-05       52
Name: last_seen, dtype: int64

The highest number of ads last seen online are on `2016-04-06`, `10772` in number, and the lowest are on `2016-03-05`, which are only `52` in number. The same frequencies presented in percentage form along with `NaN` values are as follows.

In [55]:
print(last_seen.value_counts(normalize=True, dropna=False))
print("\n")
print(last_seen.value_counts().sort_index())

2016-04-06    0.221806
2016-04-07    0.131947
2016-04-05    0.124761
2016-03-17    0.028086
2016-04-03    0.025203
2016-04-02    0.024915
2016-03-30    0.024771
2016-04-04    0.024483
2016-03-31    0.023783
2016-03-12    0.023783
2016-04-01    0.022794
2016-03-29    0.022341
2016-03-22    0.021373
2016-03-28    0.020859
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-23    0.018532
2016-03-26    0.016802
2016-03-16    0.016452
2016-03-15    0.015876
2016-03-19    0.015834
2016-03-27    0.015649
2016-03-14    0.012602
2016-03-11    0.012375
2016-03-10    0.010666
2016-03-09    0.009595
2016-03-13    0.008895
2016-03-08    0.007413
2016-03-18    0.007351
2016-03-07    0.005395
2016-03-06    0.004324
2016-03-05    0.001071
Name: last_seen, dtype: float64


2016-03-05       52
2016-03-06      210
2016-03-07      262
2016-03-08      360
2016-03-09      466
2016-03-10      518
2016-03-11      601
2016-03-12     1155
2016-03-13      432
2016

As can be seen above, 22% of the ads are last seen on `2016-04-06`. Furthermore, ads last seen online are in the time frame from `2016-03-05` to `2016-04-07`. Lastly, we shall explore the `autos["registration_year"]` column.

In [56]:
autos["registration_year"].describe()

count    48565.000000
mean      2004.755421
std         88.643887
min       1000.000000
25%       1999.000000
50%       2004.000000
75%       2008.000000
max       9999.000000
Name: registration_year, dtype: float64

The `autos["registration_year"]` column appears to be rather odd with the lowest value being year `1000` and the highest value being year `9999`.

In [57]:
# index labels denoting years arranged in an ascending order
autos["registration_year"].value_counts().sort_index().head(50)

1000     1
1001     1
1111     1
1800     2
1910     5
1927     1
1929     1
1931     1
1934     2
1937     4
1938     1
1939     1
1941     2
1943     1
1948     1
1950     3
1951     2
1952     1
1953     1
1954     2
1955     2
1956     4
1957     2
1958     4
1959     6
1960    23
1961     6
1962     4
1963     8
1964    12
1965    17
1966    22
1967    26
1968    26
1969    19
1970    38
1971    26
1972    33
1973    23
1974    24
1975    18
1976    21
1977    22
1978    44
1979    34
1980    85
1981    28
1982    41
1983    51
1984    51
Name: registration_year, dtype: int64

Any registration year before the year `1900` can be omitted safely. Also, there aren't very many registrations preceding the year `1900`. We shall now look at registration years at the tail end of this column.

In [58]:
# index labels denoting years arranged in a descending order
autos["registration_year"].value_counts().sort_index().tail(25)

2005    2936
2006    2670
2007    2277
2008    2215
2009    2085
2010    1589
2011    1623
2012    1310
2013     803
2014     663
2015     392
2016    1220
2017    1392
2018     470
2019       2
2800       1
4100       1
4500       1
4800       1
5000       4
5911       1
6200       1
8888       1
9000       1
9999       3
Name: registration_year, dtype: int64

Any year value after the year `2016` is erroneous because the year of registration of the car cannot be ahead of the year it's ad was last seen on eBay. Hence, we can safely omit values after `2016`, though there seem to be lot of values immeadiately after the year `2016`, in `2017` and `2018` but, again, for convinience's sake we shall have to omit these values from our analysis anyway.

In [59]:
# omit erroneous values
autos = autos[(autos["registration_year"] >= 1900) & (autos["registration_year"] <= 2016)]
autos["registration_year"].describe()

count    46681.000000
mean      2002.910756
std          7.185103
min       1910.000000
25%       1999.000000
50%       2003.000000
75%       2008.000000
max       2016.000000
Name: registration_year, dtype: float64

In [60]:
# display updated distribution
autos["registration_year"].value_counts(normalize=True)

2000    0.067608
2005    0.062895
1999    0.062060
2004    0.057904
2003    0.057818
          ...   
1938    0.000021
1948    0.000021
1927    0.000021
1931    0.000021
1952    0.000021
Name: registration_year, Length: 78, dtype: float64

In [61]:
# display total percentage of registration in the last 30 years
print(autos["registration_year"].value_counts(normalize=True)
                                .sort_index(ascending=False)
                                .head(30)
                                .sum()
     )

# display total percentage of registration in the last 10 years
print(autos["registration_year"].value_counts(normalize=True)
                                .sort_index(ascending=False)
                                .head(10)
                                .sum()
     )

# display total percentage of registration in the last 5 years
print(autos["registration_year"].value_counts(normalize=True)
                                .sort_index(ascending=True)
                                .head(5)
                                .sum()
     )

0.980377455495812
0.3036995779867612
0.0002142199181679913


The vast majority, i.e. 98% of the registrations have been made in the last 30 years, 30% of the registrations have been made in the last 10 years, and only 0.02% of the registrations have been made in the last 5 years.

## Exploring Price by Brand

In [62]:
# print brand column in descending order
print(autos["brand"].value_counts(normalize=True))
print('\n')
# describe brand column further
print(autos["brand"].describe())

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
mitsubishi        0.008226
honda             0.007840
kia               0.007069
alfa_romeo        0.006641
porsche           0.006127
suzuki            0.005934
chevrolet         0.005698
chrysler          0.003513
dacia             0.002635
daihatsu          0.002506
jeep              0.002271
subaru            0.002142
land_rover        0.002099
saab              0.001649
jaguar            0.001564
daewoo            0.001500
trabant           0.001392
r

The first seventeen brands shall be aggregated on in our analysis based on their respective percentages in the data set as further key:value pairs hold values too small to be considered in our current analysis.

In [63]:
# create list to store top seventeen brands
car_brands = autos["brand"].value_counts(normalize=True).head(17).index

# print top seventeen brands
print(f"Top 17 Brands: \n{list(car_brands)}")

Top 17 Brands: 
['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault', 'peugeot', 'fiat', 'seat', 'skoda', 'nissan', 'mazda', 'smart', 'citroen', 'toyota', 'hyundai']


The top twenty brands with their respective percentages are as follows:

- Volkswagen - 21.1%
- BMW - 11%
- Opel - 10.8%
- Mercedes Benz - 9.6%
- Audi - 8.7%
- Ford - 7%
- Renault - 4.7%
- Peugeot - 3%
- Fiat - 2.6%
- Seat - 1.8%
- Skoda - 1.6%
- Nissan - 1.53%
- Mazda - 1.52%
- Smart - 1.41%
- Citroen - 1.4%
- Toyota - 1.3%
- Hyundai - 1%
- Sonstige Autos - 0.98%
- Volvo - 0.91%
- Mini - 0.88%

In [64]:
# create empty dictionary to store aggregate data
brand_agg = {}

# loop through car_brands to select data
for brand in car_brands:
    selected_rows = autos[autos["brand"] == brand]
    avg_price = selected_rows["price_dollars"].mean()
    brand_agg[brand] = avg_price
    
# arrange dictionary values in a descending order
sorted_dict = dict(sorted(brand_agg.items(),
       key=lambda x: x[1],
       reverse=True)) 

print(sorted_dict)

{'audi': 9336.687453600594, 'mercedes_benz': 8628.450366422385, 'bmw': 8332.820517811953, 'skoda': 6368.0, 'volkswagen': 5402.410261610221, 'hyundai': 5365.254273504273, 'toyota': 5167.091062394604, 'nissan': 4743.40252454418, 'seat': 4397.230949589683, 'mazda': 4112.596614950635, 'citroen': 3779.1391437308866, 'ford': 3749.4695065890287, 'smart': 3580.2239031770046, 'peugeot': 3094.0172290021537, 'opel': 2975.2419354838707, 'fiat': 2813.748538011696, 'renault': 2474.8646069968195}


The most highly priced cars, on average, are made by Audi, followed by Mercedes Benz and BMW, at number two and three, after that we jump to a middle section of brands like Volkswagen, Hyundai and Nissan, and lastly we can find brands like Peugeot, Opel and Fiat which are at the less expensive end, with Renault being the least expensive among the seventeen.

## Storing Aggregate Data in a DataFrame

In [67]:
# create empty dictionary to store aggregate data
mean_mileage = {}

# loop through car_brands to select data
for brand in car_brands:
    selected_rows = autos[autos["brand"] == brand]
    avg_mileage = selected_rows["odometer_km"].mean()
    mean_mileage[brand] = avg_mileage

In [86]:
# create price series
price_series = pd.Series(sorted_dict).round(2)

# create mileage series
mileage_series = pd.Series(mean_mileage).round(2)

# create data frame
price_mileage = pd.DataFrame(price_series, columns=["mean_price"])
price_mileage["mean_mileage"] = mileage_series

price_mileage.head(6)

mean_price  mean_mileage
audi              9336.69     129157.39
mercedes_benz     8628.45     130788.36
bmw               8332.82     132572.51
skoda             6368.00     110848.56
volkswagen        5402.41     128707.16
hyundai           5365.25     106442.31

The above table makes it abundantly clear that there is no meaningful relationship among the mean prices and the mean mileages with average prices having a comparitively lower range ($5365.25 - $9336.69) and average mileage having bit higher range (106442.31km - 132572.51km).

## Conclusion

- There is no meaningful relationship between the average prices and average mileages of the top six brands.

- Germans prefer german made cars the most as the top five brands on average with regard to the number of adverts posted online are all german.

- The most expensive cars on the platform are also german with Audi, Mercedes Benz and BMW being the top three most expensive on average in the data.

- The vast majority, i.e. 98% of the registrations, for cars on the platform, have been made in the last 30 years, 30% of the registrations have been made in the last 10 years, and only 0.02% of the registrations have been made in the last 5 years.

- The highest number of ads last seen online are on `2016-04-06`, `10772` in number, and the lowest are on `2016-03-05`, which are only `52` in number.

- The highest data crawling has been done on 2016-04-03, whereas the time frame for data crawling has been from 2016-03-05 to 2016-04-07.